### PR TITLE
Filter NaN flows from MIP

### PR DESCRIPTION
**Pull request by pet-mit**
**Please check if the PR fulfills these requirements**

- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**

Bug fix

**What is the current behavior?**

- When sensitivity computation in PRAO fails for a given contingency, a sensitivity failure over-cost is computed, but the flow CNECs of that contingency are still part of the MIP range-action optimization.
This leads to NaN values in MIP (when sensi computation fails, output reference flow value is NaN) and can crash the RAO.
- When the initial sensitivity computation has a failed state, no overcost is reported

**What is the new behavior (if this is a feature change)?**
- MIP fillers that use the flow variable now filter flow CNECs first, and only retains those with a succeeded sensitivity computation status
- Unfortunately, since the status can change between two sensi iterations, we may have to remove or add variables and constraints. Since removing variables and constraints is not possible in OR-Tools, the linear problem's update method now resets the MIP and calls the "fill" method first
- PR https://github.com/powsybl/powsybl-open-rao/pull/994 has been reverted since it is no longer necessary
- An extra security has been added in OpenRaoMPSolver in order to prevent the MIP from running when there are NaN values added
- When the initial sensitivity computation has a failed state, the overcost is now reported. This means the RAO can prefer RAs that remove network divergence.
- The OpenRaoMPSolver (that uses OR-Tools under the hood) is now used in unit tests, making the tests more representative of real-time execution. However, some numerical tolerances had to be relaxed in consequence
- Also, because we have to reset MIP and call "fill" then "update" when updating now, we cannot access values used in iteration N-1 when in iteration N. Thus MaxLoopFlowFiller and MaxMinMarginFiller cannot keep values from iteration N-1 anymore (when PTDF should not be updated with PSTs). It is assumed that the updated or not values are handled outside the fillers

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [X] No

**Other information**
- This patch allows avoiding NaN values while keeping the actual fill / update APIs. 
But it does make updateBetweenSensiIterations almost useless, so we will have to refactor the code to remove all updateBetweenSensi methods and replace them with the creation of all-new problems
- It seems that overamm performance is not really impacted by the re-creation of the optimisation problem between sensi computations. But we'll have to keep an eye on the metrics.

### CODE REVIEW

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/linearproblem/OpenRaoMPSolver.java [hunk 1]
```diff
@@ -140,30 +170,65 @@ public LinearProblemStatus solve() {
...
-        return mpSolver.numVariables();
+        return variables.size();
     }
 
     public int numConstraints() {
-        return mpSolver.numConstraints();
+        return constraints.size();
+    }
+
+    public boolean minimization() {
```
Comment (phiedw) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/linearproblem/OpenRaoMPSolver.java:193]: I believe the naming convention for boolean getters is: isMinimization() / isMaximization()

**Review by phiedw — CHANGES_REQUESTED** [review]

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/fillers/MaxLoopFlowFiller.java [hunk 2]
```diff
@@ -45,18 +45,18 @@ public MaxLoopFlowFiller(Set<FlowCnec> loopFlowCnecs, FlowResult initialFlowResu
         this.loopFlowConstraintAdjustmentCoefficient = loopFlowParameters.getConstraintAdjustmentCoefficient();
     }
 
-    private Set<FlowCnec> getLoopFlowCnecs() {
-        return loopFlowCnecs;
+    private Set<FlowCnec> getValidLoopFlowCnecs(SensitivityResult sensitivityResult) {
+        return FillersUtil.getValidFlowCnecs(loopFlowCnecs, sensitivityResult);
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/fillers/MaxLoopFlowFiller.java:49]: (same question for mnecs, and also relative margin with fixed ptdfs)

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java [hunk 3]
```diff
@@ -553,6 +553,7 @@ private RaoResult runSecondPreventiveAndAutoRao(RaoInput raoInput,
                 entry.getValue().getActivatedRangeActions(entry.getKey()).forEach(rangeAction -> appliedArasAndCras.addAppliedRangeAction(entry.getKey(), rangeAction, entry.getValue().getOptimizedSetpoint(rangeAction, entry.getKey())));
             });
         // Run curative sensitivity analysis with appliedArasAndCras
+        // TODO: this is too slow, we can replace it with load-flow computations or security analysis since we don't need sensitivity values
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java:556]: I noticed it was slow for a big test case with many states with applied RAs

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/fillers/MaxLoopFlowFiller.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/fillers/MaxLoopFlowFiller.java:49]: working on it, gonna filter those too

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/linearproblem/OpenRaoMPSolver.java [hunk 4]
```diff
@@ -140,30 +170,65 @@ public LinearProblemStatus solve() {
...
+     To which we add a "bigger" double :
+       .........\\\\\\\\\\..........
+      =>
+       .........\\\\||||||..........
+       (we "lose" the least significant bits of the first double because the sum double doesn't have enough precision to show them)
+     Then we subtract the same "bigger" double:
+       .............//////..........
+       We get back our original bits for the most significant part, but the least significant bits are still gone.
+     */
+    static double roundDouble(double value, int numberOfBitsToRoundOff) {
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/linearproblem/OpenRaoMPSolver.java:224]: it's a static method. I could've put it in a new utils class but thought it wasn't worth it

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/fillers/FillersUtil.java [hunk 5]
```diff
@@ -7,23 +7,58 @@
...
+    }
+
+    /**
+     * Filters out flow CNECs that failed flow computation
+     *
+     * @param flowCnecs:  the flow CNECs to filter through
+     * @param flowResult: the flow result containing real or NaN values for CNEC flows
+     * @return a set of filtered CNECs, containing only flow CNECs with a non-NaN flow value
+     */
+    static Set<FlowCnec> getValidFlowCnecs(Set<FlowCnec> flowCnecs, FlowResult flowResult) {
```
Comment (phiedw) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/fillers/FillersUtil.java:57]: I'm not a huge fan of having two methods with the same name with arguments so similar. We could easily have a class that implements both SensitivityResult and FlowResult (eg any OptimizationResult implementation), and we wouldn't know which method to use.

**Review by phiedw — CHANGES_REQUESTED** [review]

**Review by phiedw — APPROVED** [review]

**Pull request merged**